src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -201,72 +196,91 @@ protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(N
         internal static class ElementTypeInfo<TElement>
         {
             public static readonly bool IsNonNullable =
-                typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null;
+                typeof(TElement).IsValueType && default(TElement) != null;
 
             public static readonly Type NullableElementType = IsNonNullable
                 ? typeof(Nullable<>).MakeGenericType(typeof(TElement))
                 : typeof(TElement);
         }
 
-        internal static class ArrayTypeInfo<TArrayOrList>
+        internal abstract class ArrayTypeInfo<TArray>
         {
             // ReSharper disable StaticMemberInGenericType
-            public static readonly bool IsArray;
-            public static readonly bool IsList;
-            public static readonly Type? ElementType;
-
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadArrayFunc = default!;
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<TArrayOrList>> ReadListFunc = default!;
+            public static readonly Type? ElementType = typeof(TArray).IsArray ? typeof(TArray).GetElementType() : null;
             // ReSharper restore StaticMemberInGenericType
 
-            public static bool IsArrayOrList => IsArray || IsList;
+            public static bool IsArray => ElementType is not null;
 
-            static ArrayTypeInfo()
+            static ArrayTypeInfo<TArray>? _derivedInstance;
+            static ArrayTypeInfo<TArray> DerivedInstance
             {
-                var type = typeof(TArrayOrList);
-                IsArray = type.IsArray;
-                IsList = type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>);
+                get
+                {
+                    return _derivedInstance ?? CreateInstance();
+                    static ArrayTypeInfo<TArray> CreateInstance()
+                    {
+                        if (ElementType is null)
+                            return null!;
+
+                        _derivedInstance = (ArrayTypeInfo<TArray>?)Activator.CreateInstance(typeof(ElementInfo<>).MakeGenericType(typeof(TArray), ElementType), typeof(TArray).GetArrayRank());
+                        return _derivedInstance!;
+                    }
+                }
+            }
 
-                ElementType = IsArray
-                    ? type.GetElementType()
-                    : IsList
-                        ? type.GetGenericArguments()[0]
-                        : null;
+            public static ValueTask<Array> ReadArray(ArrayHandler handler, NpgsqlReadBuffer buf, bool async, bool readAsObject = false) =>
```

Comment:
tiny nit: can you put the arrow on the next line (here and elsewhere)?

Reply:
Sure I know aspnet codebase uses the other convention... I like it more underneath as well!